src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -529,8 +529,58 @@ int G1BarrierSetC2::get_store_barrier(C2Access& access) const {
   return barriers;
 }
 
+void G1BarrierSetC2::elide_dominated_barrier(MachNode* mach) const {
+  uint8_t barrier_data = mach->barrier_data();
+  barrier_data &= ~G1C2BarrierPre;
+  if (CardTableBarrierSetC2::use_ReduceInitialCardMarks()) {
+    barrier_data &= ~G1C2BarrierPost;
+    barrier_data &= ~G1C2BarrierPostNotNull;
+  }
+  mach->set_barrier_data(barrier_data);
+}
+
+void G1BarrierSetC2::analyze_dominating_barriers() const {
+  ResourceMark rm;
+  PhaseCFG* const cfg = Compile::current()->cfg();
+
+  // Find stores and allocations, and track them in lists.
+  Node_List stores;
+  Node_List allocations;
+  for (uint i = 0; i < cfg->number_of_blocks(); ++i) {
+    const Block* const block = cfg->get_block(i);
+    for (uint j = 0; j < block->number_of_nodes(); ++j) {
+      Node* const node = block->get_node(j);
+      if (node->is_Phi()) {
+        if (BarrierSetC2::is_allocation(node)) {
+          allocations.push(node);
+        }
+        continue;
+      } else if (!node->is_Mach()) {
+        continue;
+      }
+
+      MachNode* const mach = node->as_Mach();
+      switch (mach->ideal_Opcode()) {
+      case Op_StoreP:
+      case Op_StoreN:
+        if (mach->barrier_data() != 0) {
+          stores.push(mach);
+        }
+        break;
+      default:
```

Comment:
ZGC also tries to elide barriers for atomic ops. Why isn't this done for G1 too?

Reply:
Good question. I simply assumed that it was not safe to elide barriers for atomic operations on newly allocated objects for G1, as it is not for ZGC (see motivation [here](https://github.com/openjdk/jdk/blob/5878e45ae07404bc18518e6a2cb551a7969fddd6/src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp#L764-L768)). But on second thought, that motivation seems ZGC-specific. Extending this changeset to elide G1 barriers for atomics that operate on newly allocated objects should be simple, should I do it?